### ISSUE TITLE
Disable 'misplaced-comparison-constant' checker by default by moving it to extensions

### ISSUE THREAD

**Issue by sanscore** [PyCQA/pylint#1064]
### Steps to reproduce
1.  Write a comparison such as `if 5 < a:...`
### Current behavior

Pylint complains with 'misplaced-comparison-constant'.
### Expected behavior

No issue.

In other languages, so-called ["Yoda expressions"](https://en.wikipedia.org/wiki/Yoda_conditions) can be used as a code guard to prevent an accident assignment when an equality comparison was indented. (i.e. `if (42 = x) { /* in C, this causes a compile time error error */ }` In Python, we cannot use assignments when a boolean expression is expected, so let's not dwell on this use case any further.

However, that is not my reason for using this style of comparison. I prefer to avoid using the greater-than symbol (`>`). Instead, I'll always use the less-than (`<`) or the less-than-equal (`<=`). I'll yield to Dijkstra's explanation for the best way to describe ranges in his paper ['Why numbering should start at zero'](https://www.cs.utexas.edu/users/EWD/transcriptions/EWD08xx/EWD831.html).

Given we are talking Python, that paper is, actually, very interesting because Python is one of the very few languages that the range expression is a valid boolean expression; i.e. `2 <= x < 13`. So, I'll cynically ask, "Is that half-good, half-bad?"

A quick test against pylint tells me that the range expression is handled appropriately.

So, why do we even care?

Readability!? If so, that seems inane to me. Force the nitpickers to enable `misplaced-comparison-constant` rather than ding those that aren't so easily confused.

Other rationale!? I can't think of some use case that may invoke some rare peculiarity as one may find with overloaded operators in C++. That is to say, Python will use reflection to handle equality of unlike objects; assuming the objects make use of `NotImplemented`.

Example:

``` python
"""This example is intentionally absurd."""

class MyClass(object):
    def __eq__(self, other):
        if isintance(other, int):
            return other == 42
        return NotImplemented

class MyOtherClass(object):
    def __eq__(self, other):
        return isinstance(other, MyClass)

MyClass() == 42  # True
42 == MyClass()  # True

MyOtherClass() == MyClass()  # True
MyClass() == MyOtherClass()  # True
```
### pylint --version output

```
$ pylint --version
No config file found, using default configuration
pylint 1.6.4,
astroid 1.4.7
Python 2.7.11 (default, Mar 15 2016, 09:47:58)
[GCC 4.2.1 Compatible Apple LLVM 7.0.2 (clang-700.1.81)]
```

**Comment by PCManticore** [comment]
I don't understand this issue. If this refers to the fact that misplaced-comparison-constant is enabled by default, when it should not be, I respectfully disagree. Even though Python is a language where assignments in boolean constructs are not permitted, some of its users coming from languages where this is permitted are bringing this way of writing the test into Python. It might make sense in some of the cases, but in the majority of them, it looks peculiar and pedantic.

In any way, you can either disable this locally or you can wait for Pylint 2.0, where we intend to bring multiple stages of analysis, where some of these kind of checks will be aggregated under a pedantic stage, which can be omitted entirely.

**Comment by anderson-dan-w** [comment]
Not sure I understand the reasoning to close this - there are some programmers, who come from languages where you CAN assign in boolean constructs, and maybe they shouldn't be doing that in those languages, so pylint should raise an error so they don't develop bad habits in another language?

Surely, turning this off by default wouldn't make it an error/warning to have `a > 5`; if they want to code that way they can. But it seems odd to, by default, try to enforce behavior that might apply to other languages.

**Comment by Hattes** [comment]
@anderson-dan-w I think you misunderstand. The rule being enforced by pylint is to _not_ have "5 < a", where in other languages this construction might be allowed and even encouraged because of the ability to assign in conditions.

**Comment by anderson-dan-w** [comment]
Hrm, my wording got confusing by the end, but I still stand by the general idea - I don't see any rationale as to _why_ constants shouldn't come before variables in comparisons. I read the (possible) reasoning as "even though other languages _let_ you do it, some people think that's bad practice, so let's avoid it" - which doesn't apply to Python.

If there's some other compelling reason why `5 < a` is bad, without reference to other languages or constructs that don't exist in Python, then I'm happy to reassess my view.

It's obviously not a big blocker, it's easy to disable if I feel that strongly; just wanted to reinforce the issue-opener and say I don't get why this should be the default. `pedantic` mode sounds like a good compromise, once that's available.

**Comment by PCManticore** [comment]
@anderson-dan-w I find placing a constant before a variable confusing and unnatural, just by itself and not considering the other languages where you can do assignments in conditions. Let's not forget about assignment expressions as well, which are going to land in Python 3.8, which will basically make Python a language where you can do assignments in conditions. 

As a side note, to be honest, I'm kind of tired of people complaining about a default rule or another, or a new one that I've heard this week about having agreement from Python core developers on why a rule should exist or not in `pylint`. If you don't like having it by default, `pylint` offers you the means to disable it.

**Comment by anderson-dan-w** [comment]
The 3.8 assignment expression rationale makes perfect sense, I forgot that was coming.

As for people complaining about defaults, I mean sure, fair enough. But since there isn't some set-in-stone standard about rules, it's tough to know what should and shouldn't be in there. If there are too many "defaults" that people have to change in every project, it just makes the barrier to usage that much higher.

Now, in this case, it was mentioned over 2 years ago and now I'm bringing it back up - doesn't seem to be an issue for most people, so probably not worth worrying about. But if there _are_ issues where a significant number of people think the default doesn't fit common uses, maybe things are worth revisiting.

**Comment by sanscore** [comment]
It's interesting that this issue is getting some attention. First, let me apologize for my unnecessarily long, opening comment. I don't know exactly how long that took to write, but it was too long. And, I go on abstruse tangents. 

My second regret is the title betrays my point. Which is: **This rule shouldn't exist.**

 Or, to be more diplomatic: **Why does this rule exist?**

1. Easier to read, maybe? I can't conceive of any coherent argument that would defend `x > 5` as being easier to comprehend than its opposite inequality `5 < x`; Or, `y >= 42`/ `42 <= y`, etc.
1. They're equivalent expressions, so neither solution is 'more correct' than the other.
1. In the interpreter, there are no optimizations that prefer one form over the other.

So, what's the point of having it?

 I agree that the option is fairly simple disable, so the 'issue' is mundane.

**Comment by PCManticore** [comment]
@anderson-dan-w Agreed on having thoughtful debates about defaults, we do want people to not feel overburdened with pylint's defaults.

After your and @sanscore comment, I'm reconsidering my position a bit. It seems there are some folks out there that use a lesser-or-equal constant value on the left side of the comparison intentionally, with their tools complaining about it (be either pylint or eslint). At the same time, there are other folks that just want to prohibit this idiom from their codebases (and I'm falling in this category as well). But by itself, the rule is just a convention as it's not prohibiting an actual error or faulty code that might result in some misunderstanding of its semantics, so its actual benefit is pretty low. What I think we can do, instead of completely removing it wholesale, is to mark it as a pylint extension. This means it's going to be disabled by default but folks that want to use it to prohibit the idiom from their codebase can still do that using `load-plugins`.

If anyone of you is up to that, this would be great.

**Comment by PCManticore** [comment]
I appreciate your initial comment @earonesty but please take your rhetoric somewhere else, alright? Next time you assume I'm a twitter politician you get a ban from this project.
